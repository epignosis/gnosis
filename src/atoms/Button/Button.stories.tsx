import React from "react";
import { Story } from "@storybook/react";
import Button from "./Button";

export default {
  title: "atoms/Button",
};

const Template: Story = (args) => <Button {...args} />;

export const Primary = Template.bind({});
