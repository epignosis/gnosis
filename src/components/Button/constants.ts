import { Variants } from "framer-motion";

export const iconSizes = {
  sm: 28,
  md: 30,
  lg: 32,
};

export const spinnerWrapperVariants: Variants = {
  hidden: {
    opacity: 0,
    marginInlineEnd: 6,
    transition: {
      duration: 0.1,
    },
  },
  visible: {
    opacity: 1,
    marginInlineEnd: 6,
    transition: {
      duration: 0.1,
    },
  },
};
