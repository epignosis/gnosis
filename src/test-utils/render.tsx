import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
