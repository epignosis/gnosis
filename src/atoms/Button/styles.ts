import { css, Theme, SerializedStyles } from "@emotion/react";
import { Color, Size } from "./Button";
import { hexToRGBA } from "@theme/default/colors";

const roundDimensions = {
  md: "3.5rem",
  lg: "5rem",
};

const btnHeightDimensions = {
  md: "2.5rem",
  lg: "3rem",
};

const btnPadding = {
  md: "0 1.75rem",
  lg: "0 3rem",
};

export const btnContainer = (
  { typeScaleSizes, button }: Theme,
  {
    color,
    block,
    size,
    noGutters,
  }: { color: Color; block: boolean; size: Size; noGutters: boolean },
): SerializedStyles => {
  const fontSizes = {
    md: typeScaleSizes.sm,
    lg: typeScaleSizes.lg,
  };

  return css`
    position: relative;
    font-size: ${fontSizes[size]};
    height: ${btnHeightDimensions[size]};
    min-width: ${btnHeightDimensions[size]};
    ${block ? "width: 100%" : ""};
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${button[color].default.color};
    background-color: ${button[color].default.background};
    padding: ${noGutters ? 0 : btnPadding[size]};
    border: 1px solid ${button[color].default.borderColor};
    border-radius: 5px;
    line-height: 1.125rem;
    transition: background-color 0.2s ease-in, color 0.2s ease-in, border-color 0.2s ease-in,
      box-shadow 0.2s ease;

    &:hover {
      color: ${button[color].hover.color};
      background-color: ${button[color].hover.background};
      border-color: ${button[color].hover.borderColor};

      .loading-container {
        span > span {
          background: ${button[color].hover.color};
        }
      }
    }

    &:focus {
      color: ${button[color].hover.color};
      background-color: ${button[color].hover.background};
      border-color: ${button[color].hover.borderColor};

      .loading-container {
        span > span {
          background: ${button[color].hover.color};
        }
      }
    }

    &:focus-within {
      color: ${button[color].active.color};
      background-color: ${button[color].active.background};
      border-color: ${button[color].active.borderColor};

      .loading-container {
        span > span {
          background: ${button[color].active.color};
        }
      }
    }

    &:disabled,
    &.disabled {
      color: ${button.disabled.color};
      background-color: ${button.disabled.background};
      border-color: ${button.disabled.borderColor};
      cursor: not-allowed;

      .loading-container {
        span > span {
          background: ${button.disabled.color};
        }
      }
    }

    &.rounded {
      width: ${roundDimensions[size]};
      height: ${roundDimensions[size]};
      padding: 0;
      border-radius: 50%;

      div {
        height: 100%;
        display: flex;
        align-items: center;
        justify-content: center;
      }
    }

    &.ghost {
      color: black;
      background-color: transparent;
      border: none;

      &:hover {
        color: ${button[color].ghost.color};
        background-color: ${hexToRGBA(button[color].ghost.background, 0.16)};
      }
    }

    .icon-before {
      margin-right: 0.5rem;
    }

    .icon-after {
      margin-left: 0.5rem;
    }

    .loading-container {
      span > span {
        background: ${button[color].default.color};
      }
    }

    .btn-text {
      display: inline-flex;
    }
  `;
};
