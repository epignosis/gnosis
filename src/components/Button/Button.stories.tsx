import React from "react";
import { Story } from "@storybook/react";
import Button, { Props } from "./Button";
import { CalendarSVG } from "@icons/core";

export default {
  title: "Components/Button",
  argTypes: {
    onClick: { action: "clicked" },
    color: {
      control: {
        type: "select",
        options: ["primary", "secondary", "tertiary", "danger", "success"],
      },
    },
    variant: {
      control: {
        type: "select",
        options: ["solid", "ghost"],
      },
    },
  },
  args: {
    disabled: false,
    isLoading: false,
    block: false,
    noGutters: false,
    rounded: false,
    as: "button",
    variant: "solid",
  },
};

const Template: Story<Props> = (args) => (
  <>
    <div style={{ marginBottom: 16 }}>
      <Button {...args} />
    </div>
    <div style={{ marginBottom: 16 }}>
      <Button size="lg" {...args} />
    </div>
  </>
);

export const Primary = Template.bind({});

Primary.args = {
  color: "primary",
  children: "Primary",
};

export const Secondary = Template.bind({});

Secondary.args = {
  color: "secondary",
  children: "Secondary",
};

export const Tertiary = Template.bind({});

Tertiary.args = {
  color: "tertiary",
  children: "Tertiary",
};

export const Danger = Template.bind({});

Danger.args = {
  color: "danger",
  children: "Danger",
};

export const Success = Template.bind({});

Success.args = {
  color: "success",
  children: "Success",
};

export const WithIconBefore = Template.bind({});

WithIconBefore.args = {
  color: "primary",
  iconBefore: CalendarSVG,
  children: "Icon prefix",
};

export const WithIconAfter = Template.bind({});

WithIconAfter.args = {
  color: "primary",
  iconAfter: CalendarSVG,
  children: "Icon prefix",
};
