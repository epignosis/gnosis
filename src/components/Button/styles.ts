import { css, Theme, SerializedStyles } from "@emotion/react";
import { Color, Size } from "./Button";
import { hexToRGBA } from "@theme/default/colors";

type ButtonAttrs = {
  color: Color;
  block: boolean;
  size: Size;
  noGutters: boolean;
};

const roundDimensions = {
  md: "3.5rem",
  lg: "5rem",
};

const baseButton = (
  { typeScaleSizes }: { typeScaleSizes: Theme["typeScaleSizes"] },
  { size, block, noGutters }: Omit<ButtonAttrs, "color">,
): SerializedStyles => {
  const fontSizes = {
    md: typeScaleSizes.sm,
    lg: typeScaleSizes.lg,
  };
  const btnHeightDimensions = {
    md: "2.5rem",
    lg: "3rem",
  };
  const btnPadding = {
    md: "0 1.75rem",
    lg: "0 3rem",
  };

  return css`
    position: relative;
    font-size: ${fontSizes[size]};
    height: ${btnHeightDimensions[size]};
    min-width: ${btnHeightDimensions[size]};
    ${block ? "width: 100%" : ""};
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: ${noGutters ? 0 : btnPadding[size]};
    border-radius: 5px;
    line-height: 1.125rem;
    cursor: pointer;
    transition: background-color 0.2s ease-in, color 0.2s ease-in, border-color 0.2s ease-in,
      box-shadow 0.2s ease;
  `;
};

const solidButton = (button: Theme["button"], color: Color): SerializedStyles => css`
  color: ${button[color].default.color};
  background-color: ${button[color].default.background};
  border: 1px solid ${button[color].default.borderColor};

  &:hover {
    color: ${button[color].hover.color};
    background-color: ${button[color].hover.background};
    border-color: ${button[color].hover.borderColor};

    .loading-container {
      span > span {
        background: ${button[color].hover.color};
      }
    }
  }

  &:focus {
    color: ${button[color].hover.color};
    background-color: ${button[color].hover.background};
    border-color: ${button[color].hover.borderColor};

    .loading-container {
      span > span {
        background: ${button[color].hover.color};
      }
    }
  }

  &:active {
    color: ${button[color].active.color};
    background-color: ${button[color].active.background};
    border-color: ${button[color].active.borderColor};

    .loading-container {
      span > span {
        background: ${button[color].active.color};
      }
    }
  }
`;

const outlineButton = (button: Theme["button"], color: Color): SerializedStyles => css`
  color: ${button[color].outline.color};
  background-color: transparent;
  border: 1px solid ${button[color].outline.borderColor};

  &:hover {
    color: ${button[color].hover.color};
    background-color: ${button[color].hover.background};
    border-color: ${button[color].hover.borderColor};

    .loading-container {
      span > span {
        background: ${button[color].hover.color};
      }
    }
  }

  &:focus {
    color: ${button[color].hover.color};
    background-color: ${button[color].hover.background};
    border-color: ${button[color].hover.borderColor};

    .loading-container {
      span > span {
        background: ${button[color].hover.color};
      }
    }
  }

  &:active {
    color: ${button[color].active.color};
    background-color: ${button[color].active.background};
    border-color: ${button[color].active.borderColor};

    .loading-container {
      span > span {
        background: ${button[color].active.color};
      }
    }
  }
`;

const ghostButton = (button: Theme["button"], color: Color): SerializedStyles => css`
  background-color: transparent;
  border: none;

  &:hover {
    color: ${button[color].ghost.color};
    background-color: ${hexToRGBA(button[color].ghost.background, 0.16)};
  }
`;

const linkButton = (button: Theme["button"], color: Color): SerializedStyles => css`
  color: ${button[color].link.color};
  background-color: transparent;
  border: none;

  &:hover {
    color: ${button[color].link.hoverColor};
  }
`;

const iconBeforeButton = (size: Size): SerializedStyles => {
  const btnPadding = {
    md: "0 1.75rem 0 1.25rem",
    lg: "0 3rem 0 1.875rem",
  };

  return css`
    padding: ${btnPadding[size]};

    .icon {
      margin-right: 0.5rem;
    }
  `;
};
const iconAfterButton = (size: Size): SerializedStyles => {
  const btnPadding = {
    md: "0 1.25rem 0 1.75rem",
    lg: "0 1.875rem 0 3rem",
  };

  return css`
    padding: ${btnPadding[size]};

    .icon {
      margin-left: 0.5rem;
    }
  `;
};

export const btnContainer = (
  { typeScaleSizes, button }: Theme,
  { color, size, ...attrs }: ButtonAttrs,
): SerializedStyles => {
  return css`
    ${baseButton({ typeScaleSizes }, { size, ...attrs })};

    &.solid {
      ${solidButton(button, color)};
    }

    &.outline {
      ${outlineButton(button, color)};
    }

    &.ghost {
      ${ghostButton(button, color)};
    }

    &.link {
      ${linkButton(button, color)};
    }

    &.icon-before {
      ${iconBeforeButton(size)};
    }

    &.icon-after {
      ${iconAfterButton(size)}
    }

    &:disabled,
    &.disabled {
      &,
      &:hover,
      &:focus,
      &:active {
        color: ${button.disabled.color};
        background-color: ${button.disabled.background};
        border-color: ${button.disabled.borderColor};
        cursor: not-allowed;

        .loading-container {
          span > span {
            background: ${button.disabled.color};
          }
        }
      }
    }

    &.rounded {
      width: ${roundDimensions[size]};
      height: ${roundDimensions[size]};
      padding: 0;
      border-radius: 50%;

      div {
        height: 100%;
        display: flex;
        align-items: center;
        justify-content: center;
      }
    }

    .loading-container {
      span > span {
        background: ${button[color].default.color};
      }
    }

    .btn-text {
      display: inline-flex;
    }
  `;
};
