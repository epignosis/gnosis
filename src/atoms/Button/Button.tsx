import React, { FC } from "react";
import { SerializedStyles } from "@emotion/react";
import { btnContainer } from "./styles";

const Button: FC = () => {
  return <button css={(theme): SerializedStyles => btnContainer(theme)}>Button</button>;
};

export default Button;
