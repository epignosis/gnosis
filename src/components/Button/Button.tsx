import React, { ReactElement, ReactNode, ElementType } from "react";
import { SerializedStyles } from "@emotion/react";
import { LazyMotion, domAnimation, m, AnimatePresence, Variants } from "framer-motion";
import classNames from "classnames";
import Loader from "../Loaders/Loader";
import { btnContainer } from "./styles";
import { IconType, PolymorphicComponentProps } from "types/common";

export type Color = "primary" | "secondary" | "danger" | "success";
export type Size = "md" | "lg";

export type Props = {
  color?: Color;
  size?: Size;
  noGutters?: boolean;
  block?: boolean;
  isLoading?: boolean;
  iconBefore?: IconType;
  iconAfter?: IconType;
  className?: string;
  disabled?: boolean;
  rounded?: boolean;
  children: ReactNode;
  variant?: "solid" | "outline" | "ghost" | "link";
};

export type ButtonProps<C extends ElementType> = PolymorphicComponentProps<C, Props>;

const iconSizes = {
  sm: 28,
  md: 30,
  lg: 32,
};

const spinnerWrapperVariants: Variants = {
  hidden: {
    opacity: 0,
    marginRight: 0,
    transition: {
      duration: 0.1,
    },
  },
  visible: {
    opacity: 1,
    marginRight: 6,
    transition: {
      duration: 0.1,
    },
  },
};

const Button = <C extends ElementType = "button">(props: ButtonProps<C>): ReactElement => {
  const {
    as,
    color = "primary",
    size = "md",
    variant = "solid",
    block = false,
    isLoading = false,
    noGutters = false,
    className = "",
    rounded = false,
    iconBefore,
    iconAfter,
    children,
    disabled,
    ...rest
  } = props;
  const Component = as || "button";
  const PrefixIcon = iconBefore;
  const SuffixIcon = iconAfter;
  const containerClassNames = classNames({
    [className]: Boolean(className),
    disabled: disabled || isLoading,
    rounded,
    solid: variant === "solid",
    outline: variant === "outline",
    ghost: variant === "ghost",
    link: variant === "link",
    "icon-after": SuffixIcon,
    "icon-before": PrefixIcon,
  });

  return (
    <LazyMotion features={domAnimation}>
      <Component
        css={(theme): SerializedStyles => btnContainer(theme, { color, block, size, noGutters })}
        className={containerClassNames}
        {...((as === "button" || as === undefined) && { disabled: disabled || isLoading })}
        {...rest}
      >
        <AnimatePresence>
          {isLoading && (
            <m.div
              key={props.id ?? "spinner"}
              style={{ display: "inline-flex", x: -12 }}
              className="loading"
              aria-label="loading"
              initial="hidden"
              animate="visible"
              exit="hidden"
              variants={spinnerWrapperVariants}
            >
              <Loader size="md" />
            </m.div>
          )}
        </AnimatePresence>
        {PrefixIcon && (
          <PrefixIcon height={iconSizes[size]} className="icon" data-testid="prefix-icon" />
        )}
        <span className="btn-text">{children}</span>
        {SuffixIcon && (
          <SuffixIcon height={iconSizes[size]} className="icon" data-testid="suffix-icon" />
        )}
      </Component>
    </LazyMotion>
  );
};

export default Button;
