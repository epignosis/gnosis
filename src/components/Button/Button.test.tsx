import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Button from "./Button";
import { render, screen } from "@test-utils/render";
import { CalendarSVG } from "@icons/core";

describe("<Button />", () => {
  it("renders correctly", () => {
    const btnText = faker.lorem.word();
    render(<Button type="submit">{btnText}</Button>);
    const button = screen.getByRole("button");

    expect(button).toHaveTextContent(btnText);
    expect(button).toHaveAttribute("type", "submit");
    expect(button).not.toBeDisabled();
  });

  it("calls the onClick callback", () => {
    const mockFn = jest.fn();
    render(<Button onClick={mockFn}>{faker.lorem.word()}</Button>);
    const button = screen.getByRole("button");

    expect(button).not.toBeDisabled();

    userEvent.click(button);

    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it("disabled button", () => {
    const mockFn = jest.fn();
    render(
      <Button disabled onClick={mockFn}>
        Disabled button
      </Button>,
    );
    const button = screen.getByRole("button");

    expect(button).toBeDisabled();

    userEvent.click(button);

    expect(mockFn).not.toHaveBeenCalled();
  });

  it("is loading button", () => {
    const mockFn = jest.fn();
    render(
      <Button isLoading onClick={mockFn}>
        Loading button
      </Button>,
    );
    const button = screen.getByRole("button");
    const loadingIcon = screen.getByLabelText("loading");

    expect(loadingIcon).toBeInTheDocument();
    expect(button).toBeDisabled();

    userEvent.click(button);

    expect(mockFn).not.toHaveBeenCalled();
  });

  it("act as Link button", () => {
    const linkText = faker.random.words();
    const href = faker.internet.url();

    render(
      <Button as="a" href={href}>
        {linkText}
      </Button>,
    );
    const link = screen.getByRole("link");

    expect(link).toHaveTextContent(linkText);
    expect(link).toHaveAttribute("href", href);
  });

  it("with prefix icon", () => {
    render(<Button iconBefore={CalendarSVG}>With prefix icon</Button>);
    const icon = screen.getByTestId("prefix-icon");

    expect(icon).toBeInTheDocument();
  });

  it("with suffix icon", () => {
    render(<Button iconAfter={CalendarSVG}>With suffix icon</Button>);

    const icon = screen.getByTestId("suffix-icon");
    expect(icon).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(<Button>Test button</Button>);

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with all props", () => {
    const { container } = render(
      <Button
        color="primary"
        variant="solid"
        size="lg"
        className="my-class"
        noGutters
        block
        rounded
      >
        Test button
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });
});
