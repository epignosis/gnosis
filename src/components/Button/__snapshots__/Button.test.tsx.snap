// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches snapshot 1`] = `
<div>
  <button
    class="solid css-dm6379-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
  >
    <span
      class="btn-text"
    >
      Test button
    </span>
  </button>
</div>
`;

exports[`<Button /> matches snapshot with all props 1`] = `
<div>
  <button
    class="my-class rounded solid css-1tspzsr-baseButton-baseButton-solidButton-solidButton-outlineButton-outlineButton-ghostButton-ghostButton-linkButton-linkButton-iconBeforeButton-iconBeforeButton-iconAfterButton-iconAfterButton-btnContainer-btnContainer"
    id="my-button"
  >
    <span
      class="btn-text"
    >
      Test button
    </span>
  </button>
</div>
`;
